### PR TITLE
CHE-1489: rework workspaces folder usage to fix Che on some configs

### LINKED ISSUE #1711: Docker for Windows Beta does not write workspaces and storage files to mounted drives
There seems to be a regression in the Docker for Windows Beta on Windows 10 where Che is not writing workspace and storage information to the host directory mounted to the proper location.

**Reproduction Steps:**
1. Execute Che with the following command.

```
docker run --rm=true --net=host --name=che^
 -v /home/user/che/lib:/home/user/che/lib-copy^
 -v /c/codenvy/chefile/workspaces:/home/user/che/workspaces^
 -v /c/codenvy/chefile/storage:/home/user/che/storage^
 -e "DOCKER_HOST=tcp://10.0.75.2:2375"^
 codenvy/che:nightly
```
1. Create a workspace - verify that workspace is created properly.
2. There is no data written into /c/codenvy/chefile/workspaces or /storage.  However, a new directory with the name of the workspace is written into /workspaces/<ws-name>, verifying that Che is writing into the mount.

**Expected behavior:**

Should see /storage JSON and the workspace + project files in the /workspaces.

**Observed behavior:**

Both the /workspaces and /projects directories have been created, but no files written within.  A `docker inspect` of the container verifies that the mounts are set up properly. An inspection of the `che.properties` has the storage located at ${che.home}/storage and workspaces at `${che.home}/workspaces`. `${che.home}` is verified to be `/home/user/che`.

**Che version:**    4.5.0-RC1-SNAPSHOT
**OS and version:** Windows 10  
**Docker version:** 1.12.-RC3  
**Che install:**    Docker

### PR DESCRIPTION
**Pull request by garagatyi**
### What does this PR do?

Fixes storing FS of workspaces between Che runs in case Che runs in the container. 
Removes unneeded provider.
### What issues does this PR fix or reference?

Fixes #1711 
### Previous Behavior

Che tried to create folders of projects roots of workspace on start of new workspace.
So path of projects root of workspace has to match in Che container and on docker host.
### New Behavior

It is possible to define property or environment variable to disable creation of workspace folder by Che.
In that case docker will create folders mounted to workspace container on host. This folder will have root permissions.
### Tests written?

Old tests were fixed. No new tests are added yet.
### Docs requirements?

Include the content for all the docs changes required. 
New instructions on how to run Che in container would be useful.

### CODE REVIEW

**Comment by garagatyi** [comment]
@eivantsov @TylerJewell @evoevodin @sleshchenko @mmorhun Please review

**Comment by TylerJewell** [comment]
@garagatyi - please document the property name & how it can be configured.

**Comment by TylerJewell** [comment]
@eivantsov - FYI, I am unable to build a docker image for this on a windows box right now. There seems to be some problems with docker for windows and CRLF for newlines.  Can you build an image that has this change and put it on docker hub for testing?

**Comment by mmorhun** [comment]
+1

**Comment by sleshchenko** [comment]
LGTM

**Comment by ghost** [comment]
@TylerJewell here's how you may try the fix:

`docker run --net=host -v /var/run/docker.sock:/var/run/docker.sock -v /c/Users/Tyler/che-storage:/c/Users/Tyler/che-storage -v $(pwd):/home/user/.che -e "CHE_LOCAL_CONF_DIR=/home/user/.che" -e "CHE_ALLOW_CREATION_WORKSPACE_FOLDERS=false" eivantsov/che1`

In the current directory, create che.properties file with `che.user.workspaces.storage=/c/Users/Tyler/che-storage`

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java [hunk 1]
```diff
@@ -59,68 +63,101 @@
...
+    }
+
+    @VisibleForTesting
+    @PostConstruct
+    void createNeededFolders() throws IOException {
+        if (SystemInfo.isWindows()) {
+            final Path cheHome = WindowsHostUtils.ensureCheHomeExist(createFolders);
+            final Path vfs = cheHome.resolve("vfs");
+            if (createFolders) {
+                if (!vfs.toFile().mkdir()) {
```
Comment (voievodin) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java]: There is no need to use nested `if` statement

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java [hunk 2]
```diff
@@ -59,68 +63,101 @@
...
+    public LocalWorkspaceFolderPathProvider(@Named("che.user.workspaces.storage") String workspacesMountPoint,
                                             Provider<WorkspaceManager> workspaceManager) throws IOException {
-        this.workspacesMountPoint = workspacesMountPoint;
+        this.workspacesMountPoint = getWorkspacesRootPath(workspacesMountPoint);
         this.workspaceManager = workspaceManager;
-        checkProps(workspacesMountPoint, hostProjectsFolder);
+        String allowFoldersCreationEnvVar = System.getenv(ALLOW_FOLDERS_CREATION_ENV_VARIABLE);
+        if (allowFoldersCreationEnvVar != null) {
+            if ("false".equalsIgnoreCase(allowFoldersCreationEnvVar)) {
+                createFolders = false;
```
Comment (voievodin) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java]: `createFolders = Boolean.parseBoolean(allowFoldersCreationEnvVar);`

**Comment by voievodin** [comment]
LGTM

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java (hunk 2 above)
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java]: I don't want to have this behavior. Prefer opposite.

**Comment by garagatyi** [comment]
@TylerJewell Are you OK with that PR? Can I merge it?

**Comment by TylerJewell** [comment]
@garagatyi:
1. What is the che.props entry for the same env variable?
2. Please change `CHE_ALLOW_CREATION_WORKSPACE_FOLDERS` to `CHE_CREATE_WS_FOLDERS`.

**Comment by ghost** [comment]
che.user.workspaces.storage.create_folders

**Comment by garagatyi** [comment]
Renamed `CHE_ALLOW_CREATION_WORKSPACE_FOLDERS` to `CHE_CREATE_WS_FOLDERS`

**Comment by TylerJewell** [comment]
@garagatyi @eivantsov - it failed for me.  Server was unable to boot and gave an error message in the console below.

```
docker run --net=host --rm --name=che^
 -v /c/users/tyler/che:/c/users/tyler/che^
 -v /c/codenvy:/home/user/.che^
 -e "CHE_LOCAL_CONF_DIR=/home/user/.che"^
 -e "CHE_ALLOW_CREATION_WORKSPACE_FOLDERS=false"^
 -e "DOCKER_HOST=tcp://10.0.75.2:2375"^
 eivantsov/che1
```

```
2016-07-15 11:51:05,075[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 910]        - Deploying web application archive /home/user/che/tomcat/webapps/wsmaster.war
2016-07-15 11:51:08,157[ost-startStop-1]  [WARN ] [p.DockerExtConfBindingProvider 51]   - DockerExtConfBindingProvider
2016-07-at org.eclipse.che.plugin.docker.client.InitialAuthConfig.getRegistryPrefix(InitialAuthConfig.java:99)ntext initialized event
 to listat org.eclipse.che.plugin.docker.client.InitialAuthConfig.lambda$new$0(InitialAuthConfig.java:73)
com.googat java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)rors:
        at java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1691)
1) Errorat java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)rty 'docker.registry.auth.password' is missing '.'. V
alid creat java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
  at orgat java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)java:68)
  at orgat java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)fig.java:48)
  while at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
    for at org.eclipse.che.plugin.docker.client.InitialAuthConfig.<init>(InitialAuthConfig.java:74)t>(DockerConnectorConfiguration.ja
va:123) at org.eclipse.che.plugin.docker.client.InitialAuthConfig$$FastClassByGuice$$6c93a19d.newInstance(<generated>)
  while at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:89)for the 1st parameter of org.eclipse.che.plugin.docker.client.DockerConnector.<init>(DockerConnector.java:143)
  at orgat com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:111)
```

**Comment by garagatyi** [comment]
Please send me privately your external conf file

**Comment by TylerJewell** [comment]
@garagatyi @eivantsov - Ok, got things to work after some reconfiguration. We will have to make it really clear for users that they may have to reset their docker for windows credentials to gain access otherwise they may get exception failures.

Can we please change the default che.properties file that ships to have the following comment modified.

```
# This property works for Docker for Windows, Mac, and Linux.
# On windows with virtualbox, this property is ignored.
che.user.workspaces.storage=/c/users/tyler/che/workspaces
che.user.workspaces.storage.create_folders=false
```

Can we also rename these two properties to:

```
che.workspace.storage
che.workspace.storage.create_folders
```

The run syntax that finally worked:

```
docker run --net=host --rm --name=che^
 -v /home/user/che/lib:/home/user/che/lib-copy^
 -v /c/users/tyler/che/workspaces:/c/users/tyler/che/workspaces^
 -v /c/users/tyler/che/storage:/home/user/che/storage^
 -v /c/codenvy:/home/user/.che^
 -e "CHE_LOCAL_CONF_DIR=/home/user/.che"^
 -e "CHE_ALLOW_CREATION_WORKSPACE_FOLDERS=false"^
 -e "DOCKER_HOST=tcp://10.0.75.2:2375"^
 eivantsov/che1
```

**Comment by garagatyi** [comment]
@TylerJewell here is current comment

```
### Storage and user configuration
# The location where your workspaces (and their projects) are stored.
# This property is ignored on Windows due to limitations with VirtualBox
# On Windows, all workspaces stored in c:\%userprofile%\AppData\Local\Eclipse Che\
che.user.workspaces.storage=${che.home}/workspaces
```

Do you want to have next comment instead?

```
### Storage and user configuration
# The location where your workspaces (and their projects) are stored.
# This property works for Docker for Windows, Mac, and Linux.
# On windows with virtualbox, this property is ignored. Instead of that all workspaces stored in c:\%userprofile%\AppData\Local\Eclipse Che\
```

**Comment by garagatyi** [comment]
If we rename property **che.user.workspaces.storage** then users with configured external property files will miss the projects after upgrade. Are you OK with that?

**Comment by TylerJewell** [comment]
@garagatyi 

```
### Storage and user configuration
# The location where your workspaces (and their projects) are stored.
# This property works for Docker for Windows, Mac, and Linux.
# On virtualbox for windows, this property is ignored. Instead, they are stored at:
# c:\%userprofile%\AppData\Local\Eclipse Che\
```

**Comment by TylerJewell** [comment]
@garagatyi - yes, ok with the breaking change.  Is there anyway we can quietly support the old configuration in our code?  So that there are two properties for the same value?  We will only document and show the new one but quietly honor the old?

**Comment by garagatyi** [comment]
> We will only document and show the new one but quietly honor the old?

Yes, it is possible

**Comment by TylerJewell** [comment]
@garagatyi - ok, after changes of comments and property names, +1 from me - ok to merge.

**Comment by garagatyi** [comment]
@TylerJewell What then do if both old and new properties are set?

**Comment by TylerJewell** [comment]
@garagatyi - accept new one, I think.

**Comment by garagatyi** [comment]
Looks like we will have to accept the old one because it will be the case when user overrided old property and then upgraded Che.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java [hunk 3]
```diff
@@ -59,68 +66,118 @@
...
     @Inject
-    public LocalWorkspaceFolderPathProvider(@Named("host.workspaces.root") String workspacesMountPoint,
+    public LocalWorkspaceFolderPathProvider(@Named("che.workspace.storage") String workspacesMountPoint,
                                             Provider<WorkspaceManager> workspaceManager) throws IOException {
         this.workspacesMountPoint = workspacesMountPoint;
         this.workspaceManager = workspaceManager;
-        checkProps(workspacesMountPoint, hostProjectsFolder);
+        this.isWindows = SystemInfo.isWindows();
+        String allowFoldersCreationEnvVar = System.getenv(ALLOW_FOLDERS_CREATION_ENV_VARIABLE);
+        if (allowFoldersCreationEnvVar != null) {
```
Comment (sleshchenko) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/local/node/provider/LocalWorkspaceFolderPathProvider.java]: You can remove this check. Or merge with next by &&
I prefer first variant

**Comment by sleshchenko** [comment]
LGTM

**Comment by TylerJewell** [comment]
@garagatyi - did this improvement remove the need for the volume mount?  If so, once eugene creates a new image, I can test.  Otherwise LGTM.

**Comment by garagatyi** [comment]
Yes. You don't have to mount workspaces folder into Che container if folders creation is disabled (e.g. in your case).

**Comment by TylerJewell** [comment]
@garagatyi - can we add an environment variable for `che.workspace.storage` property named `CHE_WORKSPACE_STORAGE`?  By having both properties available as environment variables, we can configure both of these with -e commands without the user having to create a custom che.properties file. So right now, Windows users will always have to create a custom properties file and we should give them environment variables only.

Also, to start the process of more consistency - I updated the name of the properties and the environment variables that match.  Sorry for moving this around.  I had to write some docs on this and get a feel before I could settle in on a standard.

```
che.workspace.storage
che.workspace.storage.create_folders
```

```
CHE_WORKSPACE_STORAGE
CHE_WORKSPACE_STORAGE_CREATE_FOLDERS
```

**Comment by TylerJewell** [comment]
Documentation update:
https://eclipse-che.readme.io/v4.4/docs/usage-docker-for-windows

**Comment by garagatyi** [comment]
Now mounting /workspaces folder into Che container is not needed. storage path can be set with env variable

**Pull request merged**